timer/timer.go

```diff
@@ -1,50 +1,100 @@
+// a scheduled timeout is triggered
 type Timeout struct {
 	Height process.Height
 	Round  process.Round
 }
 
+// SizeHint implements surge SizeHinter for Timeout
+func (timeout Timeout) SizeHint() int {
+	return surge.SizeHint(timeout.Height) +
+		surge.SizeHint(timeout.Round)
+}
+
+// Marshal implements surge Marshaler for Timeout
+func (timeout Timeout) Marshal(buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := surge.Marshal(timeout.Height, buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("marshaling Height=%v: %v", timeout.Height, err)
+	}
+	buf, rem, err = surge.Marshal(timeout.Round, buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("marshaling Round=%v: %v", timeout.Round, err)
+	}
+
+	return buf, rem, nil
+}
+
+// Unmarshal implements surge Unmarshaler for Timeout
+func (timeout *Timeout) Unmarshal(buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := surge.Unmarshal(&timeout.Height, buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("unmarshaling Height: %v", err)
+	}
+	buf, rem, err = surge.Unmarshal(&timeout.Round, buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("unmarshaling Round: %v", err)
+	}
+
+	return buf, rem, nil
+}
+
+// LinearTimer defines a timer that implements a timing out functionality.
+// The timeouts for different contexts (Propose, Prevote and Precommit) are
+// provided as callback functions that handle the corresponding timeouts. The
+// timeout scales linearly with the consensus round
 type LinearTimer struct {
-	opts               Options
-	onTimeoutPropose   chan<- Timeout
-	onTimeoutPrevote   chan<- Timeout
-	onTimeoutPrecommit chan<- Timeout
+	opts                   Options
+	handleTimeoutPropose   func(Timeout)
+	handleTimeoutPrevote   func(Timeout)
+	handleTimeoutPrecommit func(Timeout)
 }
 
-func NewLinearTimer(opts Options, onTimeoutPropose, onTimeoutPrevote, onTimeoutPrecommit chan<- Timeout) process.Timer {
+// NewLinearTimer constructs a new Linear Timer from the input options and channels
+func NewLinearTimer(opts Options, handleTimeoutPropose, handleTimeoutPrevote, handleTimeoutPrecommit func(Timeout)) process.Timer {
 	return &LinearTimer{
-		opts:               opts,
-		onTimeoutPropose:   onTimeoutPropose,
-		onTimeoutPrevote:   onTimeoutPrevote,
-		onTimeoutPrecommit: onTimeoutPrecommit,
+		opts:                   opts,
+		handleTimeoutPropose:   handleTimeoutPropose,
+		handleTimeoutPrevote:   handleTimeoutPrevote,
+		handleTimeoutPrecommit: handleTimeoutPrecommit,
 	}
 }
 
+// TimeoutPropose schedules a propose timeout with a timeout period appropriately
+// calculated for the consensus height and round
 func (t *LinearTimer) TimeoutPropose(height process.Height, round process.Round) {
 	go func() {
 		time.Sleep(t.timeoutDuration(height, round))
-		t.onTimeoutPropose <- Timeout{Height: height, Round: round}
+		t.handleTimeoutPropose(Timeout{Height: height, Round: round})
```

Comment:
Allow the handler to be nil.
```suggestion
		if t.handleTimeoutPropose != nil {
			t.handleTimeoutPropose(Timeout{Height: height, Round: round})
		}
```

Reply:
In that case I will simply wrap the entire `go func() { ... }()` by the `nil` check? If the handler is `nil`, then we can probably avoid the `time.Sleep` as well?